src/main/scala/mesosphere/marathon/state/AppDefinition.scala

```diff
@@ -495,13 +507,18 @@ object AppDefinition {
     appDef.executor should matchRegexFully("^(//cmd)|(/?[^/]+(/[^/]+)*)|$")
     appDef is containsCmdArgsOrContainer
     appDef.healthChecks is every(portIndexIsValid(appDef.portIndices))
+    appDef.readinessChecks is every(ReadinessCheck.readinessCheckValidator(appDef.portDefinitions))
     appDef.instances should be >= 0
     appDef.fetch is every(fetchUriIsValid)
     appDef.mem should be >= 0.0
     appDef.cpus should be >= 0.0
     appDef.instances should be >= 0
     appDef.disk should be >= 0.0
-    appDef must definesCorrectResidencyCombination
+    appDef must complyWithResidencyRules
+    appDef must complyWithSchedulerLabelRules
+    appDef must complyWithSingleInstanceLabelRules
+    appDef must complyWithReadinessCheckRules
+    (appDef.isScheduler is false) or (appDef.upgradeStrategy is UpgradeStrategy.validForSchedulers)
```

Comment:
Since there are 2 rules regarding the upgradeStrategy, I moved them together so this reads
`appDef must complyWithUpgradeStrategyRules`